Add Java/Kotlin interop docs (#4557)
dart-lang/site-www: src/_guides/libraries/java-interop.md

---
title: "Java interop using package:jnigen"
description: "To use Java in your Dart program, use package:jnigen."
jnigen: "https://pub.dev/packages/jnigen"
jni: "https://pub.dev/packages/jni"
example: "https://github.com/HosseinYousefi/jnigen_example/tree/main"

> jnidoc: "https://docs.oracle.com/javase/7/docs/technotes/guides/jni/spec/jniTOC.html"

Reviewer: Consider linking to the version of this for the latest JDK LTS, at least if jnigen supports it:

```suggestion
jnidoc: "https://docs.oracle.com/en/java/javase/17/docs/specs/jni/index.html"
```

---

Dart mobile, command-line, and server apps

> running on the [Dart Native platform](/overview#platform), on Android, Windows

Reviewer: Is macOS not supported yet?

Reviewer 2: Looks like it's not https://github.com/dart-lang/jnigen#supported-platforms

Reviewer: Ah, thanks for the link! Seems `macOS` is supported, just not for Flutter yet according to the table there. Would be nice to include macOS, since many Dart developers use that as their development platform. If it's important that Flutter doesn't support it, we could add a Flutter note or a simple exception in parentheses.

and Linux can use [`package:jni`]({{page.jni}})
and [`package:jnigen`]({{page.jnigen}})
to call Java and Kotlin APIs.

{{site.alert.note}}
  This interop feature is **experimental**,

>   and [in active development](https://github.com/dart-lang/jnigen/issues).

Reviewer: Consider linking to an overarching issue such as https://github.com/dart-lang/sdk/issues/49674 so developer can follow the progress easier.

{{site.alert.end}}

`package:jni` allows Dart code to interact
with Java through [JNI]({{page.jnidoc}}). 
However, doing so involves a lot of boilerplate code,
so you can use `package:jnigen` to automatically generate
the Dart bindings for a given Java API.

You can decompile Kotlin to Java, allowing `package:jnigen`
to generate bindings for Kotlin as well.

> ## Simple Java Example

Reviewer 2: Sentence case to match other headings

```suggestion
## Simple Java example
```

This guide walks you through [an example]({{page.example}})
that uses `package:jnigen` to generate bindings for a simple class
that can show a toast.

> Along with JDK, maven is required. 

Reviewer: Is there a version requirement for the JDK used?

Author: No

> Along with JDK, maven is required. 

Reviewer: Consider cross-linking to Maven documentation or installation instructions here: https://maven.apache.org/index.html

More and more Java developers don't use Maven directly these days so may not be aware of what it is.

Reviewer 2: ```suggestion
Along with JDK, [Maven][] is required. 

[Maven]: https://maven.apache.org/

```

> It's recommended to have `clang-format` installed to format the generated
> C bindings.

Reviewer: Consider putting this into a tip:

```suggestion
{{site.alert.ti}}
  It's recommended to have `clang-format` installed to
  format the generated C bindings.
{{site.alert.end}}
```

Also consider cross-linking to more information about `clang-format` as well.

Reviewer 2: I think the best way to do this would be a "Prerequisites" section, for this line, and the JDK/Maven line. And link to each

> ### Simple Java Example
> This guide walks you through [an example](https://github.com/dart-lang/site-www/blob/main/src/_guides/libraries/%7B%7Bpage.example%7D%7D) that uses `package:jnigen` to generate bindings for a simple class that can show a toast.
> 
> #### Prerequisites
> - [JDK]()
> - [Maven]()
> - (Optional) [`clang-format`]() to format the generated C bindings
> 
> ### Configuring jnigen

Reviewer: Love it!

> ### Configuring jnigen

Reviewer 2: ```suggestion
### Configuring `jnigen`
```

First, add `package:jni` as a dependency and `package:jnigen`

> as a dev dependency.

Reviewer 2: Good to cross-link on uncommon terms
```suggestion
as a [dev dependency][].

[dev dependency]: /tools/pub/dependencies#dev-dependencies
```

```terminal

> $ dart pub add jni
> $ dart pub add --dev jnigen

Reviewer 2: `pub add` has [new recommended syntax](https://dart.dev/tools/pub/cmd/pub-add#dev-dependency)! So at minimum this needs to be written like:
 ```suggestion
$ dart pub add jni
$ dart pub add dev:jnigen
```

But you could also add them in the same line:
```suggestion
$ dart pub add jni dev:jnigen
```

```

Next, create a top-level file called `jnigen.yaml`. This will contain
the configuration for generating the bindings.

```yaml
output:
  c:
    library_name: example
    path: src/example/
  dart:
    path: lib/example.dart
    structure: single_file

source_path:
  - 'java/' 
classes:
  - 'dev.dart.Example'
```

`path` specifies the path for the generated `c` and `dart` bindings.

> `source_path` specifies the path of the java source file that we want to

Reviewer 2: ```suggestion
`source_path` specifies the path of the Java source file that we want to
```

generate bindings for, and `classes` specifies the Java class.

`java/dev/dart/Example.java` contains a very simple class, which has a public
static method called `sum`:

```java
package dev.dart;

public class Example {
  public static int sum(int a, int b) {
    return a + b;
  }
}
```

### Generating the Dart bindings

To generate the Dart (and C) bindings, run jnigen and specify the config file
using the `--config` option.

```terminal
$ dart run jnigen --config jnigen.yaml
```

For this example, this will generate
[lib/example.dart]({{page.example}}/lib/example.dart), just as we specified in
`jnigen.yaml`.

This file contains a class called `Example`, which has a static method called
`sum`, just like the Java file.

### Using the bindings

Now you're ready to load and interact with the generated library.
The example app, [bin/sum.dart]({{page.example}}/bin/sum.dart), gets two numbers
as arguments and prints their sum. Using the `Example.sum` method is identical
to Java.

```dart

> // a and b are integer arguments passed.

Reviewer 2: Did you want to say "a and b are integer arguments passed to ...."? The Java class? 
If not you can just say:
```suggestion
// a and b are integer arguments
```

print(Example.sum(a, b));
```

### Running the example

> Before running the example, you must build the dynamic libraries for jni and the

Reviewer 2: ```suggestion
Before running the example, you must build the dynamic libraries for `jni` and the
```

---
title: "Java interop using package:jnigen"
description: "To use Java in your Dart program, use package:jnigen."
jnigen: "https://pub.dev/packages/jnigen"
jni: "https://pub.dev/packages/jni"
example: "https://github.com/HosseinYousefi/jnigen_example/tree/main"
jnidoc: "https://docs.oracle.com/en/java/javase/17/docs/specs/jni/index.html"
---

Dart mobile, command-line, and server apps
running on the [Dart Native platform](/overview#platform), on Android, Windows,
macOS, and Linux can use [`package:jni`]({{page.jni}})
and [`package:jnigen`]({{page.jnigen}})
to call Java and Kotlin APIs.

{{site.alert.note}}
  This interop feature is **experimental**,
  and [in active development](https://github.com/dart-lang/sdk/issues/49674).
{{site.alert.end}}

`package:jni` allows Dart code to interact
with Java through [JNI]({{page.jnidoc}}). 
However, doing so involves a lot of boilerplate code,
so you can use `package:jnigen` to automatically generate
the Dart bindings for a given Java API.

You can decompile Kotlin to Java, allowing `package:jnigen`

> to generate bindings for Kotlin as well.

Author: It can be used on both Java byte code and Java source. Kotlin needs to be first compiled to Java byte code to be usable by jnigen.

I plan to add that example later.

> You can decompile Kotlin to Java, allowing `package:jnigen`
> to generate bindings for Kotlin as well.

Reviewer: Sounds good, thanks!

Should this say the following then?

```suggestion
You can compile Kotlin to Java bytecode, allowing `package:jnigen`
to generate bindings for Kotlin as well.
```

## Simple Java example

This guide walks you through [an example]({{page.example}})
that uses `package:jnigen` to generate bindings for a simple class

> that can show a toast.

Reviewer 2: What does this mean? I feel like I'm missing something obvious 😅

Author: I was first creating a Flutter example for Android, which showed a toast 😂 I left this part after editing it to be a Dart standalone example.

jnidoc: 'https://docs.oracle.com/en/java/javase/17/docs/specs/jni/index.html'
---

Dart mobile, command-line, and server apps
running on the [Dart Native platform](/overview#platform), on Android, Windows,
macOS, and Linux can use [`package:jni`]({{page.jni}})
and [`package:jnigen`]({{page.jnigen}})
to call Java and Kotlin APIs.

{{site.alert.note}}
This interop feature is **experimental**,
and [in active development](https://github.com/dart-lang/sdk/issues/49674).
{{site.alert.end}}

`package:jni` allows Dart code to interact
with Java through [JNI]({{page.jnidoc}}).
However, doing so involves a lot of boilerplate code,
so you can use `package:jnigen` to automatically generate
the Dart bindings for a given Java API.

You can compile Kotlin to Java bytecode, allowing `package:jnigen`
to generate bindings for Kotlin as well.

## Examples

The following examples show how to use `package:jnigen`:

| **Example**             | **Description**                                                                                   |
| ----------------------- | ------------------------------------------------------------------------------------------------- |

> | [in_app_java][]         | Demonstrates how to include custom Java code in Flutter application and call that using `jnigen`. |

Reviewer 2: ```suggestion
| [in_app_java][]         | Demonstrates how to include custom Java code in a Flutter application and call that using `jnigen`. |
```
